workqueue: add delayed metric

huww98 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Use synctest and get rid of wait loop.

Reporting the number of items currently waiting for retry. This metric is suitable for an alert if it is above 0 for a long time.

Previous retry metrics is not suitable for alerting because it is hard to distinguish a random transient error from a backoff repeating error.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Workqueue gains a new "delayed_items" metric, reporting the number of items waiting for retry.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

huww98 committed:
workqueue: add delayed metric

Reporting the number of items currently waiting for retry. This metric is suitable for an alert if it is above 0 for a long time.

Previous retry metrics is not suitable for alerting because it is hard to distinguish a random transit error from a backoff repeating error.

siyuanfoundation commented:
/cc @richabanker for metrics review

dgrisonnet commented on staging/src/k8s.io/component-base/metrics/prometheus/workqueue/metrics.go:
```diff
@@ -93,8 +93,15 @@ var (
 
+	delayed = k8smetrics.NewGaugeVec(&k8smetrics.GaugeOpts{
+		Subsystem:      WorkQueueSubsystem,
+		Name:           "delayed",
```
In terms of naming I would suggest:
- "delayed_items"
- "delayed_queue_depth"
- "delayed_entries"

dgrisonnet commented:
> I'm not sure I got your point. What your said is exactly the reason we are not satisfied with the existing retries metrics. We have many errors which are fixed automatically by retry within 50ms, and we don't care. For these retries, the proposed delayed metric will only go up transiently, most likely not even be captured by Prometheus.

The problem is that alerting on the size of the delayed metric being > 0 is bound to be noisy. Even if there isn't a 1000s item in it, there is a very high chance that whenever Prometheus will scrape this metric, there will be at least one delayed element.
I don't know precisely what's done around observability in this part of the code, but I think producing a log for items that have been delayed for a long time is reasonable and would be a good way to identify bad actors.

> it cannot be distinguished from the normal retires. But the proposed delayed metric will keep above zero.

Yes that to me is a problem if you try to create an alert where you can't distinguish if it is a false positive or not. Maybe introducing a histogram capturing the p50/90/99 of delay would be better for solve your problem?

huww98 commented:
/reopen
/remove-lifecycle rotten
@dgrisonnet Hi, are you still interested in reviewing this? We faced another issue where our ratelimiter holds 400k items back and delays the workqueue for hours. That gave us a headache to diagnosis why our items are missing. It would be much easier with this metric in place.

dgrisonnet commented:
@huww98 sorry for the delay, I missed this PR.

Could you reduce the scope of the changes to only adding the new metric? Right now there are a lot of unrelated changes

huww98 commented:
The changes are splited into 2 commits, one covert the tests to use synctest, the other add the new metric.

But OK, if you want, I will drop the first commit.

dgrisonnet commented:
From your first commit, I don't think the changes to the clock from `staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go` are required for the new metric

huww98 committed:
workqueue: drop unnecessary blank receiver names

The `_` blank identifier on noopMetricsProvider method receivers carries
no meaning; omit it.

huww98 commented:
Dropped unrelated changes. I will open them as followup cleanup PRs afterwards

> From your first commit, I don't think the changes to the clock from staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go are required for the new metric

Reverted now. They are not required. Rationale of that change:
- keep `nextReadyAt` at nil to save an allocation
- "there is no reason to prefer NewTimer when After will do."  https://pkg.go.dev/time#After
